javascript/ql/lib/semmle/javascript/ApiGraphs.qll

```diff
@@ -699,6 +774,98 @@ module API {
           lbl = Label::parameter(1) and
           ref = awaited(call)
         )
+        or
+        decoratorDualEdge(base, lbl, ref)
+        or
+        decoratorUseEdge(base, lbl, ref)
+        or
+        // for fields and accessors, mark the reads as use-nodes
+        decoratorPropEdge(base, lbl, ref.(DataFlow::PropRead))
+      )
+    }
+
+    /** Holds if `base` is a use-node that flows to the decorator expression of the given decorator. */
+    pragma[nomagic]
+    private predicate useNodeFlowsToDecorator(TApiNode base, Decorator decorator) {
+      exists(DataFlow::SourceNode decoratorSrc |
+        use(base, decoratorSrc) and
+        trackUseNode(decoratorSrc).flowsToExpr(decorator.getExpression())
+      )
+    }
+
+    /**
+     * Holds if `ref` is both a use and def-node that should an incoming edge from `base` labelled `lbl`.
+     *
+     * This happens because the decorated value escapes into the decorator function, and is then replaced
+     * by the function's return value. In the JS analysis we generally assume decorators return their input,
+     * but library models may want to find the return value.
+     */
+    private predicate decoratorDualEdge(TApiNode base, Label::ApiLabel lbl, DataFlow::Node ref) {
+      exists(ClassDefinition cls |
+        useNodeFlowsToDecorator(base, cls.getADecorator()) and
+        lbl = Label::decoratedClass() and
+        ref = DataFlow::valueNode(cls)
+      )
+      or
+      exists(MethodDefinition method |
+        useNodeFlowsToDecorator(base, method.getADecorator()) and
+        not method instanceof AccessorMethodDefinition and
+        lbl = Label::decoratedMember() and
+        ref = DataFlow::valueNode(method.getBody())
+      )
+      or
+      exists(Parameter param |
+        useNodeFlowsToDecorator(base, param.getADecorator()) and
+        lbl = Label::decoratedParameter() and
+        ref = DataFlow::parameterNode(param)
```

Comment:
I'm unsure if there should be a def-node for this parameter-node.   
And the tests pass just fine if I move the `decoratedParameter` edge to the `decoratorUseEdge` predicate.

Reply:
It's technically possible for the parameter decorator to replace the whole method with one that treats that parameter as a sink.

I have never seen any decorators that would have such behaviour, though. A made-up example might be a decorator that wraps some expensive runtime checks on a value, which could then be a DoS sink.

In any case, it turns out using a parameter node as a sink is actually a terrible idea, at least with the current implementation (see test case in [ca145f2](https://github.com/github/codeql/pull/8576/commits/ca145f21b04534312047c78eecbf19bae82c332f)).

So I've changed it to only generate use-nodes 👍